packages/medusa/src/api/routes/store/carts/create-cart.ts

```diff
@@ -186,19 +204,64 @@ export default async (req, res) => {
       const createdCart = await cartServiceTx.create(toCreate)
 
       if (validated.items?.length) {
-        const generateInputData = validated.items.map((item) => {
-          return {
-            variantId: item.variant_id,
-            quantity: item.quantity,
-          }
-        })
-        const generatedLineItems: LineItem[] = await lineItemServiceTx.generate(
-          generateInputData,
-          {
-            region_id: regionId,
-            customer_id: req.user?.customer_id,
-          }
-        )
+        let generatedLineItems: LineItem[]
+
+        if (
+          featureFlagRouter.isFeatureEnabled(
+            IsolateProductDomainFeatureFlag.key
+          )
+        ) {
+          const variantIds = validated.items
+            .map((i) => i.variant_id)
+            .filter(Boolean)
+
+          const remoteQuery = req.scope.resolve("remoteQuery")
+
+          const variants = await retrieveVariantsWithIsolatedProductModule(
+            remoteQuery,
+            variantIds
+          )
+
+          validateItemsInput(validated.items, variants)
+
+          const variantsIdsMap = new Map(variants.map((v) => [v.id, v]))
+
+          const generateInputData = validated.items.map((item) => {
+            const variant = variantsIdsMap.get(item.variant_id)
+
+            if (!variant) {
+              throw new MedusaError(
+                MedusaError.Types.INVALID_DATA,
+                `Variant with id: ${item.variant_id} not found`
+              )
+            }
```

Comment:
todo: I believe it is handle as part of `validateItemsInput`, should we remove it?